readme.md
# Ungrad fork of the elexis-3-core Repository

## Build:

Prerequisites:

* Java 8
* Maven > 3.2

```
git clone https://github.com/rgwch/elexis-3-core
cd elexis-3-core
./build.sh
```
## Download of precompiled binaries:

<http://www.elexis.ch/ungrad2018/products/elexis-core/latest/>

## Jenkins Build history

<http://www.elexis.ch:8080/>

## Upgrade database

If upgrading from ungrad (2016), read [here](http://www.elexis.ch/ungrad/reference/update_34/) (or try your luck after making a backup - it might run automatically).

If upgrading/transgrading from Legacy-Elexis: Apply the SQL Script `ch.elexis.core.data - rsc/dbScripts/apply_ungrad2018_db.sql`
This should not break compatibility. It just adds some ungrad specific extensions. Applying it several times also won't hurt.

## Main differences to elexis/elexis-3-core

* Modified Login Dialog

* Case-insensitive usernames

* User with appropiate privileges can always modify case Details (Falldetails) 

* modified extinfo fields

* medication ui:
  * a bit more responsive
  * more user feed back on long running operations
  
* Artikelliste: 
   * Blackbox-Filter
   * substance filter
   * ability to import only pharma or non-pharma articles or both.
   * display of contianed substances for pharma-articles.
   * Replace "Mepha Sorter" with a more generic "Preferred Provider Sorter"
  
* Database
    * ID fields changed to 36 Chars (RFC 4122 UUIDv4)
    * Time-Field in "Behandlungen" - allows for better sorting od same-day consultations
    * time field in Behandlungen.

## See also

<http://www.elexis.ch/ungrad>

<http://github.com/rgwch/elexis-3-base>

<http://github.com/rgwch/elexis-ungrad>


Changelog
# Generated by update_changelog.rb on 2018.03.09
# similar to git log --date=iso --pretty=format:'%H %ad %an %s' release/3.0.25..release/3.1.0

410 commits between release/3.2.0b30 (2017.06.16) and release/3.5.0 (2018.03.10)
--------------------------------------------------------------------------------
   number changes by authors are
   140: Marco Descher
   107: thomashu
   100: sazgin
    58: Niklaus Giger
     2: Txomin Astrain
     2: DavidGutknecht
     1: rgwch
--------------------------------------------------------------------------------
bddc66a139765215a344e3b1225152414332807f 2018-03-08 16:55:26 +0100        Niklaus Giger [10984] Use apostrophes for ORIGIN_BRANCH
a456ef80cfa5fa421a0f21b2b25989d8a96f9920 2018-03-08 16:44:58 +0100        Niklaus Giger [10984] Update p2-site
44e214d08b45336107eef37b869bc39837843e80 2018-03-08 14:08:18 +0100        Niklaus Giger [10984] Update .gitlab-ci and update_changelog.rb
a53d2d4083ccd1221db3fa949ae424eebdfd35c7 2018-02-15 10:40:56 +0100        Niklaus Giger [9664] Create mediawiki changelog for all repositories
9dfa3e4f49052a73b0e9ad5054af712ce7cbbaaf 2018-03-08 09:29:26 +0100        Marco Descher [10984] Update productive target
5da57269d0cba2dbfcbefba6573e3ffb79029029 2018-03-08 09:21:13 +0100        Marco Descher [10984] Create Branch 3.5
f9658b6f73b107eb9395a77db440778dfe48f60b 2018-03-07 13:23:54 +0100        Marco Descher [10962] Incorrect pathologic mark
ee749e6b6651a9b9ac88d59dcc2b6f2a64cfe3f0 2018-03-06 09:02:58 +0100        Marco Descher [10510] Re-introduce MAT import on HL7ReaderV25
bc813b7cdbb0dae3934243748243e212dd128a5d 2018-03-05 17:33:05 +0100             thomashu [10931] added comparison methods with resolution parameter
59a8b27525ffd50ee5d2097a276a484b908ebc39 2018-03-03 15:06:31 +0100             thomashu [10931] reset TimeTool resolution for comparision, added warning comment
4b02bc98cfa3016fe596c8c96568e546c2e37404 2018-03-02 13:29:54 +0100        Marco Descher [8913] Enhance StatusUtil, ObjectStatus
e9445869f0a4a36c05757a362f86164db2749147 2018-03-01 08:30:28 +0100       DavidGutknecht Fehler korrigiert in der vorherige Anpassung (#242)
49b89cd299772a1be706b48ec2454906b2105e8c 2018-02-23 10:22:19 +0100        Marco Descher [10441] Show warning on setting preference useLocalLabRefValue
0e6717b44c13c5da12aeca0a48ee4555a42569ec 2018-02-23 07:57:15 +0100       DavidGutknecht Try to use mimetype field to decide with which Program a Document wil… (#240)
0bdff8e5dba4c4fec5ebd81449e1b92ffc8eec34 2018-02-22 14:57:54 +0100             thomashu [9080] added usage statistics plugin to core udpate site
c51e50fade2117ea09146f0b3014f45520afe1e2 2018-02-22 14:46:02 +0100             thomashu [9080] added some comments
29fff7f7257c259f8042fe1d85528b567e5d7fc3 2018-02-22 13:08:29 +0100             thomashu [10843] pass enabled state to sub control of LabeledInputField
e7c31482a15c1d39a6ce1103e7b95fae868c5d69 2018-02-20 14:44:22 +0100             thomashu [10843] disable editing of patnr on kontakt detail of patient
d132f14d306492afa8fd7efc08b6a591e18ed027 2018-02-15 14:43:02 +0100        Marco Descher [10740] Re-arrange laboratory preference pages
047c7c8c07ce0804e85ce2377fcedaa3e04ab073 2018-02-15 13:03:10 +0100             thomashu [10719] shutdown ElexisEventDispatcher before disposing the Display
f0421855d1ede60a419c8fca5bb8f93184441831 2018-02-15 10:36:52 +0100             thomashu [10719] added null and disposed checks in async runnables
541717f2bdea01cef3df03542fdce3d795b0c6db 2018-02-14 17:01:05 +0100             thomashu [10719] added null and disposed checks in async runnables
c76dc1d0a06a3189339da1bdb3251a5ac4022644 2018-02-14 16:09:32 +0100             thomashu [10785] use optifier context methods to change side in case of exclusion
663cd84a9dce8d9e59389108c552c5fb6b1727e5 2018-02-14 16:08:03 +0100             thomashu [10785] add context methods to IOptifier interface
0642e0a058424964b4120e3eb053ea999d426398 2018-02-14 14:32:10 +0100        Marco Descher [10796] Move Kontakt.XID definitions to XIDConstants
864a64adda3df78c3136864eae51157e9e27ec4b 2018-02-14 12:11:14 +0100             thomashu [10718] changed tolerance for getting active medication from 5 to 10 sec
da23b1bfa156adca3ed3782e43ae6febba61ab55 2018-02-13 16:20:21 +0100             thomashu [10521] make GlobalEventDispatcher more thread safe
3769ae5e1c30b91aedd54a8907b57a063fa94183 2018-02-13 15:44:33 +0100             thomashu [10521] removed IElexisEventDispatcher, clean up ElexisEventDispatcher
754d4de6b997ed55ba23cc3c9224826cec56c821 2018-02-13 15:45:00 +0100        Marco Descher [10655] Fix tests
3e7f219619dc966f026dabb475992b7e8b882139 2018-02-13 15:34:47 +0100        Marco Descher [10655] Fix tests
a632a09d3974a035cdae6cfb2e9bc423e822e919 2018-02-13 14:39:32 +0100             thomashu [10719] added null and disposed checks in async runnables
e41f3814824990fe76a2e8077f7573227259741b 2018-02-13 10:11:58 +0100        Marco Descher [10663] missing translation
8c16705888a430b58622a8a5f2cc926d80849247 2018-02-13 09:48:53 +0100        Marco Descher [10663] mark indetermined pathologic flags with yellow background
0295e934f855d6fe473c068e51cc284b755befc7 2018-02-13 08:10:00 +0100        Marco Descher [10786] Fix isPathologic determination for hl7 text values
d7328108825873e8d4ee89f10f9d50c848986e38 2018-02-12 13:12:15 +0100        Marco Descher [10740] Multiple laboratory refactorings
0107611819c4b973f729bff38807d9968d9f6b0d 2018-02-12 11:02:47 +0100        Marco Descher [9956] Index out of bounds at CTabFolder.createItem
e56dc263ddf269f2d2082ea4d7e5634166db8623 2018-02-09 16:59:07 +0100        Niklaus Giger [9664] Update index.html for Artikelstamm.elexis.info
55eea932e99a9bc62a908f516438f316528ada34 2018-02-09 08:52:28 +0100        Marco Descher [9771] Missing translation in ReminderView
1970a8deab634e49b534ed091583fb0a69aa2977 2018-02-07 15:03:10 +0100        Marco Descher [10655] Refactor HL7v25-ORUR01 import, include absence of flag

ch.elexis.core.releng/update_changelog.rb
#!/usr/bin/env ruby

require 'pp'
begin
require 'pry'
rescue LoadError
end
require 'ostruct'
require 'rugged'
require 'trollop'
require 'oga'
require 'open-uri'
require 'yaml'
require 'csv'
require 'fileutils'
require 'active_support/all'

ISSUE_URL = 'https://redmine.medelexis.ch/issues'
CACHE_FILE = File.join(Dir.home, '.cache/redmine_issues.yaml')
IGNORE_ISSUE_STATUS = /Abgewiesen|erledigt|zur*ckgestell/i
@options = Trollop::options do
  version "#{File.basename(__FILE__)} (c) 2107 by Niklaus Giger <niklaus.giger@member.fsf.org>"
  banner <<-EOS
Useage:
  if --from is given, --to must be given, too, e.g. --from=release/3.0.25 --to=release/3.1.0
  Collects all release tag, limits them to the local history, creates statistic for
  each tag and then the Changelog
  The issues fetched from #{ISSUE_URL} are cached
  in #{CACHE_FILE} as loading tickets takes a non negligeable time.
  Issues with a status not matching #{IGNORE_ISSUE_STATUS} are reloaded if
  they were not already update via the Redmine-API today.
  #{version}
EOS

  opt :force_tag,       "Use HEAD and force it as tag_name and use it a to tag", :type => String, :default => nil
  opt :with_tickets,    "Emit also information from Medelexis Redmine (needs API) ",  :default => false
  opt :changelog,       'Name of file to be written', :type => String, :default => 'Changelog'
  opt :from,            'Only a difference from the given tag', :type => String, :default => nil
  opt :to,              'Only a difference to the given tag', :type => String, :default => nil
  opt :mediawiki,       'Create Changelog.mediawiki with a summary grouped by weekyl changes/ticket'
end

require 'rugged'
MAX_ID = 999
FORCE_TAG_NUMERIC = (MAX_ID.to_s*4).to_i
DATE_FORMAT = '%Y.%m.%d'
MEDIAWIKI_FILE = 'Changelog.mediawiki'
Issue = Struct.new(:id, :subject, :fixed_version, :git_version,  :status, :project, :last_api_fetch)
CommitInfo = Struct.new(:ticket, :author_date, :committer_date, :text)
class CommitInfo
  def commit_week
    date = Date.parse(committer_date)
    sprintf('%04d Woche %02d', date.year, date.cweek)
  end
end
@scriptStarted = Time.now
@options[:with_tickets] = true if @options[:mediawiki]
if @options[:with_tickets]
  @csv_file_name = "#{@options[:changelog]}.csv"
end
@options[:to] = @options[:force_tag] if  @options[:force_tag]

def tag_name_to_numerical_value(tag_name)
  items = []
  if tag_name.index('.b') # beta_vesion
    # release/3.2.0.beta22 0> [3, 2, 0, 22]
    items = tag_name.split('/').last.split(/\.(beta|b)|\./).collect{|x| x.to_i if /^\d/.match(x) }.compact
  else
    items = tag_name.split('/').last.split('.').collect{|x| x.to_i}
    items.insert(-2, MAX_ID)
  end
  sprintf('%03i%03i%03i%03i', items[0], items[1], items[2], items[3]).to_i
end

TAG_INFO = Struct.new('TAG_INFO', :tag_name, :numerical, :tag, :commit_id, :parent, :tag_date)
@all_taginfos = []

def get_taginfo_by_name(tag_name)
  @all_taginfos.find{ |taginfo| taginfo.tag_name.eql?(tag_name)}
end

def get_release_tags
  @repo = Rugged::Repository.new(Dir.pwd)
  tags = @repo.tags.find_all{|x| x.name.index('release/')}
  tags_hash = {}
  tags.each do |tag|
    next unless /^release/i.match(tag.name)
    next if /alpha/i.match(tag.name)
    tags_hash[tag.name] = tag.target_id
    @all_taginfos << TAG_INFO.new(tag.name, tag_name_to_numerical_value(tag.name),
                                  tag, tag.target_id, nil,
                                  tag.target.committer[:time].strftime(DATE_FORMAT))
  end
  if @options[:force_tag]
    info =  TAG_INFO.new(@options[:force_tag], FORCE_TAG_NUMERIC, @repo.head, @repo.head.target_id)
    info.commit_id =  @repo.head.target_id
    info.tag_date = (Date.today+1).strftime(DATE_FORMAT)
    @all_taginfos << info
  end
  tags_hash
rescue => error
  puts error
  puts error.backtrace.join("\n")
end

def find_tags_in_current_branch
  newer_id = @repo.references['HEAD'].target_id
  newer_id = @repo.head.target_id
  walker = Rugged::Walker.new(@repo)
  walker.sorting(Rugged::SORT_TOPO | Rugged::SORT_REVERSE) # optional
  walker.push(newer_id)
  res = walker.collect{|c| c};
  @tag_in_local_branch = []
  res.each do |commit|
    if (info = @all_taginfos.find{|taginfo| taginfo.commit_id == commit.oid})
      @tag_in_local_branch << info
      puts "Adding local tag #{info.tag_name} commit #{info.commit_id}"  if $VERBOSE
    end
  end
end

def build_branch_history
  previous = nil
  sorted = @tag_in_local_branch.sort_by { |info| info.numerical}
  puts "Sorted are #{sorted.collect{ |x| x.tag_name}.join(',')}"  if $VERBOSE
  sorted.each_with_index do |current, idx|
    unless previous
      previous = current
      next
    end
    @history = current
    current.parent = previous.clone
    previous = current.clone
  end
end

def get_history(old_id, newer_id, git_version)
  old = @repo.lookup(old_id)
  newer = @repo.lookup(newer_id)
  return get_history(newer_id, old_id, git_version) if old.time > newer.time

  info = OpenStruct.new
  walker = Rugged::Walker.new(@repo)
  # walker.sorting(Rugged::SORT_TOPO || Rugged::SORT_DATE) # optional
  walker.push(newer_id)
  walker.hide(old_id)

  info.commits = []
  info.authors = {}
  walker.each do |commit|
    author = commit.author[:name]
    ticket_id = (m =  /^\[(\d*)\]/.match(commit.message)) && m[1]
    ticket = ticket_id ? read_issue(ticket_id) : Issue.new
    ticket.git_version = git_version if ticket
    line = "#{commit.oid} #{commit.author[:time]} #{sprintf('%20s', author)} #{commit.message.split("\n").first}"
    info.commits  << CommitInfo.new(ticket, commit.author[:time].strftime(DATE_FORMAT), commit.committer[:time].strftime(DATE_FORMAT), line)
    info.authors[author] ||= 0
    info.authors[author] += 1
  end
  info
end


def read_issue(id = 5760)
  @nr_loaded ||= 0
  unless ENV['REDMINE_MEDEXIS_API']
    puts "Environment variable REDMINE_MEDEXIS_API must be specified to read an issue from the Medelexis Redmine"
    raise "Missing REDMINE_MEDEXIS_API variable"
  end
  if (ti = @ticket_cache[id])
    ti.last_api_fetch ||= Date.today
    if IGNORE_ISSUE_STATUS.match(ti.status) || ti.last_api_fetch = Date.today
      puts "Skip reading ticket #{id} with status #{ti.status} dated #{ti.last_api_fetch}" if $VERBOSE
      return ti
    end
  end
  issue = Issue.new
  issue.last_api_fetch = Date.today
  content =  open("#{ISSUE_URL}/#{id}.xml",
      "User-Agent" => "Ruby/#{RUBY_VERSION}",
      "X-Redmine-API-Key" => "#{ENV['REDMINE_MEDEXIS_API']}"
    ).read
  document = Oga.parse_xml( content ) ;
  ['id', 'subject'].each do |field|
    cmd = "issue.#{field} = document.xpath('issue/#{field}').first.text"
    eval(cmd)
  end
  ['project', 'status', 'fixed_version'].each do |field|
    cmd = "value = document.xpath('issue/#{field}').first; issue.#{field} = value ? value.get('name') : nil"
    eval(cmd)
  end
  issue.subject = issue.subject[0..79] if issue.subject
  puts "Fetched ticket #{id} #{issue.status}" if $VERBOSE
  @nr_loaded += 1
  $stdout.write "\n (re-)loaded #{@nr_loaded}/#{@ticket_cache.size} tickets." if @nr_loaded % 100 == 0
  $stdout.write '.'
  @ticket_cache[id] = issue
  issue
rescue OpenURI::HTTPError => error
  puts "Issue #{id} not found"
  return nil
rescue => error
  puts error
  puts error.backtrace.join("\n")
end

def emit_mediawiki_changelog(all_commits, file)
  mediawiki = File.open(file, 'w+')
  mediawiki.puts "= Changelog für Elexis 3.5 ab 3.5.0 ="
  mediawiki.puts ""
  mediawiki.puts "Die Zahlen in Klammern beziehen sich auf das nicht öffentlich zugängliche Ticket-System der Firma Medelexis AG"
  mediawiki.puts ""
  sorted = all_commits.sort{|left, right| left.committer_date <=> right.committer_date}
  by_week = {}
  sorted.reverse.each do |commit|
    next unless commit.ticket && commit.ticket.id
    by_week[commit.commit_week] ||= {}
    by_week[commit.commit_week] [commit.ticket.id] = commit
    by_week[commit.commit_week] [:sunday] = Date.parse(commit.committer_date).sunday
  end
  by_week.each do | week, values|
    ids = values.keys.find_all{|x| x.is_a?(String)}.uniq.sort;
    mediawiki.puts "==  #{values[:sunday].strftime('%Y.%m.%d')}: #{ids.size} gelöste Tickets =="
    mediawiki.puts ""
    ids.each do |id|
      ti = values[id].ticket
      line = "* #{sprintf("%14s", "'''([https://redmine.medelexis.ch/issues/" + ti.id + ' ' + ti.id + "])'''")} #{ti.subject.gsub(/\n|\r\n/, ',').strip}"
      mediawiki.puts line
    end
    mediawiki.puts ""
  end
end

def emit_changes(ausgabe, from=nil, to=nil)
  raise "from and to must be given" unless to && from
  from_tag =@all_taginfos.find{|x| x.tag_name == from}
  to_tag =@all_taginfos.find{|x| x.tag_name == to}
  info = get_history(from_tag.commit_id, to_tag.commit_id, to_tag.tag_name)
  header = []
  header << ''
  line = "#{info.commits.size} commits between #{from_tag.tag_name} (#{from_tag.tag_date}) and #{to_tag.tag_name} (#{to_tag.tag_date})"
  header << line
  header << '-' * line.size
  header << '   number changes by authors are'
  info.authors.sort_by{|key, value| value}.reverse.each do |author, nr_commits|
    header << "   #{sprintf("%3i", nr_commits)}: #{author}"
  end
  header << '-' * line.size

  header << ''
  puts header if $VERBOSE
  ausgabe.puts header.join("\n")
  ausgabe.puts info.commits.collect{|info| info.text}.join("\n")
  if @csv_file_name
    unless @csv_file # create header
      @csv_file = CSV.open(@csv_file_name, "wb", :encoding => 'utf-8')
      @csv_file << ['id', 'status', 'fixed_version', 'git_version','author_date', 'committer_date', 'subject', 'project', 'text']
    end
    puts "\nAdding #{sprintf('%-25s', to_tag.tag_name)}" #  to #{@csv_file_name}"
    info.commits.each do |ci|
      next unless ci.ticket && ci.ticket.id.to_i > 0
      ti = ci.ticket
      line = [ti.id, ti.status, ti.fixed_version, ti.git_version, ci.author_date, ci.committer_date,
              ti.subject.gsub(/\n|\r\n/, ',').strip, ti.project, ci.text.gsub(/\n|\r\n/, ' ').strip]
      @csv_file << line
    end
  end
end

def emit_history(filename, from = nil, to = nil)
  walk = @history
  filename += "-#{from.sub('release/','')}-#{to.sub('release/','')}" if from && to
  File.open(filename, 'w+') do |ausgabe|
    ausgabe.puts "# Generated by #{File.basename(__FILE__)} on #{Time.now.strftime(DATE_FORMAT)}"
    ausgabe.puts "# similar to git log --date=iso --pretty=format:'%H %ad %an %s' release/3.0.25..release/3.1.0"
    if from && to
      emit_changes(ausgabe, from, to)
    else
      while walk && walk.parent
        unless @tag_in_local_branch.find{|x| x.commit_id == walk.parent.commit_id}
          walk = walk.parent
          next
        end
        emit_changes(ausgabe, walk.parent.tag_name, walk.tag_name)
        walk = walk.parent
      end
    end
  end
end

@ticket_cache = File.exist?(CACHE_FILE) ? YAML.load_file(CACHE_FILE) : {}
@ticket_cache = {} unless @ticket_cache.is_a?(Hash)
@nr_loaded ||= 0
$stdout.sync = true
puts "Loaded #{@ticket_cache.size} entries from #{CACHE_FILE}"
at_exit do
  puts "Saving #{@ticket_cache.size} entries to #{CACHE_FILE}"
  FileUtils.mv(CACHE_FILE, CACHE_FILE+ '.backup', :verbose => true) if File.exist?(CACHE_FILE) && File.size(CACHE_FILE) > 100
  File.open(CACHE_FILE,'w+') do |h|
    h.write @ticket_cache.to_yaml
  end if @nr_loaded.size != @ticket_cache.size
end
# Order of next calls is necessary!
<<<<<<< HEAD
@history = []
=======
>>>>>>> 3.5
if @options[:mediawiki]
  raise "from and to must be given" unless @options[:from] && @options[:to]
  git_directories = (Dir.glob('.git') + Dir.glob('*/.git')).collect{|x| File.expand_path(File.dirname(x)) }
  mediawiki_file = File.join(Dir.pwd, MEDIAWIKI_FILE)
  history_all = []
  git_directories.each do |repo|
    Dir.chdir(repo)
    @all_taginfos = []
    @history = []
    get_release_tags
    from_tag =@all_taginfos.find{|x| x.tag_name ==  @options[:from] }
    unless from_tag
      puts "Skipping #{repo} as there is no tag #{@options[:from]}"
      next
    end
    to_tag =@all_taginfos.find{|x| x.tag_name ==  @options[:to]}
    info = get_history(from_tag.commit_id, to_tag.commit_id, to_tag.tag_name)
    history_all +=info.commits
    puts "  #{repo}: found #{info.commits.size} history_all now #{history_all.size}"
  end
  emit_mediawiki_changelog(history_all, mediawiki_file)
  puts "Created #{mediawiki_file} #{File.size(mediawiki_file)} bytes"
else
  get_release_tags
  if @options[:from] && @options[:to]
    emit_history(@options[:changelog], @options[:from], @options[:to])
  else
    find_tags_in_current_branch
    build_branch_history
    emit_history(@options[:changelog])
  end
end
puts "\nCreated #{File.expand_path(@options[:changelog])}" +
    "#{@csv_file_name ? ' and ' +  File.expand_path(@csv_file_name) : ' '}" +
    "for #{@options[:from] ? @options[:from] : 'first commit'} up to #{@options[:to] ? @options[:to] : 'HEAD'}"
@scriptStopped = Time.now
@diffSeconds = (@scriptStopped-@scriptStarted).to_i
puts "#{Time.now}: Script finished after #{sprintf('%i:%02i', @diffSeconds/60, @diffSeconds%60)}. Reloaded #{@nr_loaded} redmine tickets"
